torch/csrc/distributed/rpc/rref_context.h

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include <c10/util/Optional.h>
+#include <torch/csrc/distributed/rpc/message.h>
+#include <torch/csrc/distributed/rpc/rpc_agent.h>
+#include <torch/csrc/distributed/rpc/rref.h>
+#include <torch/csrc/distributed/rpc/types.h>
+
+#include <atomic>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+
+// Manages RRef lifetime and keeps track of RRef forks.
+class RRefContext {
+ public:
+  static void initInstance(std::shared_ptr<RpcAgent>);
+  static std::unique_ptr<RRefContext>& getInstance();
+
+  RRefContext(const RRefContext&) = delete;
+  void operator=(const RRefContext&) = delete;
+
+  worker_id_t getWorkerId() const;
+  RRefId genRRefId();
+  const std::shared_ptr<RpcAgent>& agent() const;
+
+  // create a new RRef
+  template <typename T>
+  std::shared_ptr<OwnerRRef<T>> createOwnerRRef(worker_id_t ownerId) {
+    TORCH_CHECK(ownerId == getWorkerId(), "Cannot create OwnerRRef on user.");
+    return getOrCreateOwnerRRef<T>(genRRefId());
+  }
+
+  std::shared_ptr<UserRRef> createUserRRef(worker_id_t ownerId) {
+    TORCH_CHECK(ownerId != getWorkerId(), "Cannot create UserRRef on owner.");
+    return createUserRRef(ownerId, genRRefId(), genRRefId());
+  }
+
+  std::shared_ptr<UserRRef> createUserRRef(
+      worker_id_t ownerId,
+      const RRefId& rrefId,
+      const ForkId& forkId) {
+    TORCH_CHECK(
+        ownerId != getWorkerId(), "RRef owner cannot create user RRef.");
+    // RRefContext does not track user RRefs, it will be destructed when there
+    // is no shared_ptrs pointing to it. NB: cannot use make_shared here as the
+    // constructor of UserRRef is private
+    return std::shared_ptr<UserRRef>(new UserRRef(ownerId, rrefId, forkId));
+  }
+
+  // get an existing RRef or create a new one from a serialized
+  // ``RRefForkData``.
+  template <typename T>
+  std::shared_ptr<RRef> getOrCreateRRef(at::IValue&& value) {
+    auto rfd = RRefForkData::fromIValue(std::move(value));
+    return getOrCreateRRef<T>(rfd.ownerId_, rfd.rrefId_, rfd.forkId_);
+  }
+
+  template <typename T>
+  std::shared_ptr<RRef> getOrCreateRRef(
+      worker_id_t ownerId,
+      const RRefId& rrefId,
+      const ForkId& forkId) {
+    if (ownerId == getWorkerId()) {
+      return getOrCreateOwnerRRef<T>(rrefId);
+    } else {
+      return createUserRRef(ownerId, rrefId, forkId);
```

Comment:
should we separate out getOrCreateOwnerRRef() and getOrCreateUserRRef()? Do we need getOrCreateUserRRef?